src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -1464,127 +1660,85 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
-                            break;
-                        case TypeCode.String:
-                            if (TryGenericSort(keys as string[], items, index, length)) return;
-                            break;
-                        case TypeCode.UInt16:
-                            if (TryGenericSort(keys as ushort[], items, index, length)) return;
-                            break;
-                        case TypeCode.UInt32:
-                            if (TryGenericSort(keys as uint[], items, index, length)) return;
-                            break;
-                        case TypeCode.UInt64:
-                            if (TryGenericSort(keys as ulong[], items, index, length)) return;
-                            break;
-                    }
+                    case TypeCode.Boolean:
+                        if (TryGenericSort<bool>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.Byte:
+                        if (TryGenericSort<byte>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.Char:
+                        if (TryGenericSort<char>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.Double:
+                        if (TryGenericSort<double>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.Int16:
+                        if (TryGenericSort<short>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.Int32:
+                        if (TryGenericSort<int>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.Int64:
+                        if (TryGenericSort<long>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.SByte:
+                        if (TryGenericSort<sbyte>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.Single:
+                        if (TryGenericSort<float>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.UInt16:
+                        if (TryGenericSort<ushort>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.UInt32:
+                        if (TryGenericSort<uint>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.UInt64:
+                        if (TryGenericSort<ulong>(keys, items, index, length)) return;
+                        break;
+                }
 
-                    static bool TryGenericSort<TKey>(TKey[]? keys, Array? items, int index, int length)
+                static bool TryGenericSort<T>(Array keys, Array? items, int index, int length) where T : struct
```

Comment:
The original FCalls was optimize only for the case where keys and items are the same type. I do not think we need to have a qudratic generic expansion here to do better.

Reply:
I hadn't noticed that the FCall required them to be of the same type.  Happy for the simplification.